include/kernel.h

```diff
@@ -1977,6 +1977,30 @@ static inline bool k_queue_remove(struct k_queue *queue, void *data)
 	return sys_sflist_find_and_remove(&queue->data_q, (sys_sfnode_t *)data);
 }
 
+/**
+ * @brief Append an element to a queue only if it's not present already.
+ *
+ * This routine appends data item to @a queue. The first 32 bits of the
+ * data item are reserved for the kernel's use. Appending elements to k_queue
+ * relies on sys_slist_is_node_in_list which is not a constant time operation.
+ *
+ * @note Can be called by ISRs
+ *
+ * @param queue Address of the queue.
+ * @param data Address of the data item.
+ *
+ * @return true if data item was added, false if not
+ */
+static inline bool k_queue_unique_append(struct k_queue *queue, void *data)
+{
+	if (sys_sflist_is_node_in_list(&queue->data_q,
+			(sys_sfnode_t *)data) == false) {
+		k_queue_append(queue, data);
+		return true;
+	}
```

Comment:
Can't you put the FOREACH search right here?  It wouldn't be meaningfully longer or more complicated.

Reply:
Thanks for the review @andyross . Yes I could have directly implemented `k_queue_unique_append` using `SYS_SLIST_FOR_EACH_NODE`. I thought a "contains" test will be useful and hence implemented this way. But I also have the patch ready with implementation as you suggested here. I am waiting for @andrewboie's review on my PR